app/javascript/app/document-capture/components/form-steps.jsx

```diff
@@ -1,16 +1,68 @@
-import React, { useState } from 'react';
+import React, { useEffect, useState } from 'react';
 import PropTypes from 'prop-types';
 import Button from './button';
 import useI18n from '../hooks/use-i18n';
 import useHistoryParam from '../hooks/use-history-param';
 
+/**
+ * @typedef FormStep
+ *
+ * @prop {string}                    name      Step name, used in history parameter.
+ * @prop {import('react').Component} component Step component implementation.
+ * @prop {(values:object)=>boolean}  isValid   Step validity function. Given set of form values,
+ *                                             returns true if values satisfy requirements.
+ */
+
+/**
+ * Given a step object and current set of form values, returns true if the form values would satisfy
+ * the validity requirements of the step.
+ *
+ * @param {FormStep} step   Form step.
+ * @param {object}   values Current form values.
+ */
+export function isStepValid(step, values) {
+  const { isValid = () => true } = step;
+  return isValid(values);
+}
+
+/**
+ * Hook which ensures that given the current set of steps and form values, the current step state is
+ * valid per the requirements of the step. At mount, if any step is not satisfied by the current
+ * form values, the current step will be set to that step.
+ *
+ * @param {FormStep[]}              steps       Form steps.
+ * @param {object}                  values      Form values.
+ * @param {FormStep|undefined}      currentStep Current step, if known.
+ * @param {(nextStep:string)=>void} setStepName Step setter.
+ */
+function useVerifiedCompletion(steps, values, currentStep, setStepName) {
+  useEffect(() => {
+    if (!currentStep) {
+      return;
+    }
+
+    for (let i = 0; i < steps.length; i += 1) {
+      const step = steps[i];
+      if (step.name === currentStep.name) {
+        break;
+      }
+
+      if (!isStepValid(step, values)) {
+        setStepName(step.name);
```

Comment:
what if we had this function return a step name if it's valid, and then put the control flow, side effects back in the caller?

```
let nextStepName = useVerifiedCompletion(steps, values, step);
if (nextStepName) {
  setStepName(nextStepName);
}
```

Reply:
> what if we had this function return a step name if it's valid, and then put the control flow, side effects back in the caller?
> 
> ```
> let nextStepName = useVerifiedCompletion(steps, values, step);
> if (nextStepName) {
>   setStepName(nextStepName);
> }
> ```

I can see how it's a bit cleaner to have the extracted logic be responsible primarily for determining the furthest step which satisfies validation. But since `setStepName` is a side effect and (aside from hooks) the component render should remain pure, we'd at the very least still need to contain this within a `useEffect`. Part of the goal of extracting a hook here was to have this be a little more self-documenting by function name and hook documentation, because there'd otherwise be nothing to stop us from putting the entire `useEffect` within `FormSteps`. It'd certainly help avoid all the ugly arguments passing. If we're reintroducing `useEffect`, it kinda puts us back toward that part of the balance of value in extracting any utility function here.

I'll want to think on it more (and I welcome feedback), but I could see some compromise solution which combines utility of "get the last valid step" and reintroducing the effect back into the main component logic:

```js
const stepIndex = stepName ? steps.findIndex((_step) => _step.name === stepName) : 0;
const step = steps[stepIndex];
useEffect(() => {
  const lastValidStepIndex = getLastValidStepIndex(steps, values);
  if (lastValidStepIndex < stepIndex) {
    setStepIndex(steps[lastValidStepIndex].name);
  }
}, []);
```

Reply:
How to handle side effects is definitely more your wheelhouse than mine, so I'm to defer to you. I was able to trace the existing code, so it's not indecipherable as is, so if you think it's the best fit, happy to leave it as is! This was just my outsider's view of the React code.

Reply:
I pushed some revisions in 0eee895. In some ways, it's a bit more complex in how it manages computing the "effective" step and syncing it, but there's also a few advantages:

- Rather than render the _wrong_ step and correct it only after the initial mount of the component (could potentially cause a brief flicker for the user), we always render the step which should be rendered based on validity.
- In this regard, there's a consolidation of logic, since considering the absence of a history parameter as prompting the first step is a similar concern.
- It's easier to unit test methods like `getStepIndexByName` and `getLastValidStepIndex`, vs. testing a React hook like `useVerifiedCompletion`.